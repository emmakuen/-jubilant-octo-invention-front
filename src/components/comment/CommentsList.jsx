import { Comment, CommentReplyDialog } from "..";
import useDialog from "../../hooks/useDialog";

const CommentsList = ({ comments, user }) => {
  const { openDialog, closeDialog, checkIsDialogOpen } = useDialog();
  const commentProps = { openDialog, user };

  return (
    <div data-testid="comment-list">
      {comments.map((comment) => {
        const dialogProps = { user, closeDialog, parentId: comment.id };
        const isDialogOpen = checkIsDialogOpen(comment.id);
        const hasReplies = !!comment.replies;
        return (
          <div key={comment.id}>
            <Comment
              {...commentProps}
              isDialogOpen={isDialogOpen}
              comment={comment}
              isParent={!!comment.replies}
            />
            {hasReplies &&
              comment.replies.map((reply) => (
                <Comment
                  key={reply.id}
                  {...commentProps}
                  parentId={comment.id}
                  isDialogOpen={isDialogOpen}
                  comment={reply}
                  isReply
                />
              ))}
            {isDialogOpen && <CommentReplyDialog {...dialogProps} />}
          </div>
        );
      })}
    </div>
  );
};

export default CommentsList;
