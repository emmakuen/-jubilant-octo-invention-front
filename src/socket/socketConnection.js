import { io } from "socket.io-client";
import * as commentHelpers from "../helpers/comment";

const UPVOTE_RECEIVED_EVENT = "comment-upvote-updated";
const UPVOTE_EMIT_EVENT = "upvote";

let socket;

export const connectWithSocketServer = () => {
  socket = io(process.env.REACT_APP_API_URL);

  socket.on("connect", () => {
    // console.log(`Connected to socket.io server with id ${socket.id}...`);
  });
};

export const getSocketInstance = () => {
  return socket;
};

export const emitToggleUpvote = ({ commentId, userId }) => {
  socket.emit(UPVOTE_EMIT_EVENT, { commentId, userId });
};

export const listenForToggleUpvote = (commentId, userId, setCurrentComment) => {
  socket.on(UPVOTE_RECEIVED_EVENT, (upvotedComment) => {
    if (upvotedComment.id !== commentId) return;
    const populatedComment = commentHelpers.populateUpvotedComment(
      upvotedComment,
      userId
    );

    setCurrentComment(populatedComment);
  });
};

export const removeToggleUpvoteListener = () => {
  socket.off(UPVOTE_RECEIVED_EVENT, listenForToggleUpvote);
};
