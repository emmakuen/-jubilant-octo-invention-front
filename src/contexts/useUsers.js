import {
  useContext,
  createContext,
  useState,
  useEffect,
  useCallback,
  useMemo,
} from "react";
import * as userService from "../services/user";
import useLocalStorage from "../hooks/useLocalStorage";
import * as socketConnection from "../socket/socketConnection";

const LOCAL_STORAGE_KEY = "comment-ui-user";

export const UsersContext = createContext();

export const UsersProvider = ({ children }) => {
  const [users, setUsers] = useState([]);
  const [user, setUser] = useLocalStorage(LOCAL_STORAGE_KEY, null);
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState(null);

  useEffect(() => {
    const _fetchUsers = async () => {
      setLoading(true);
      try {
        socketConnection.connectWithSocketServer();
        const usersData = await userService.getUsers();
        setUsers(usersData);
      } catch (err) {
        setError(err);
        console.error(err);
      }
      setLoading(false);
    };

    _fetchUsers();
  }, []);

  const login = useCallback((userData) => setUser(userData), [setUser]);

  const value = useMemo(
    () => ({ users, user, loading, login, error }),
    [error, loading, login, user, users]
  );

  return (
    <UsersContext.Provider value={value}>{children}</UsersContext.Provider>
  );
};

export const useUsers = () => useContext(UsersContext);
