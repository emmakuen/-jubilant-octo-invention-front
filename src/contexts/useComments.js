import { createContext, useContext, useState, useCallback } from "react";
import * as commentService from "../services/comment";
import * as commentHelpers from "../helpers/comment";
import { useUsers } from "./useUsers";
import * as commentLikeService from "../services/commentLike";
import * as socketConnection from "../socket/socketConnection";

export const CommentsContext = createContext();

export const CommentsProvider = ({ children }) => {
  const [comments, setComments] = useState([]);
  const [error, setError] = useState(null);
  const { user } = useUsers();

  const postComment = async ({ content, parentId = null }) => {
    try {
      const postedComment = await commentService.postComment({
        content,
        authorId: user.id,
        parentId,
      });
      const populatedComment = commentHelpers.populatePostedComment(
        postedComment,
        user
      );

      if (!parentId) {
        // add the posted comment first to show it in descending order
        setComments([populatedComment, ...comments]);
      } else {
        const updatedComments = commentHelpers.updateComments(
          comments,
          parentId,
          populatedComment
        );
        setComments(updatedComments);
      }
    } catch (err) {
      console.error(err);
      setError(err);
    }
  };

  const fetchComments = useCallback(async () => {
    try {
      const commentsData = await commentService.getParentComments();
      const repliesData = await commentService.getAllRepliesGroupedByParentId();
      const populatedComments = commentHelpers.populateComments(
        commentsData,
        repliesData,
        user?.id
      );
      setComments(populatedComments);
      console.log(populatedComments);
    } catch (err) {
      console.error(err);
      setError(err);
    }
  }, [user?.id]);

  const toggleCommentLike = async (commentId) => {
    if (!commentId) return;
    try {
      await commentLikeService.toggleCommentLike({
        commentId,
        userId: user.id,
      });

      // update in realtime about the upvote
      socketConnection.emitToggleUpvote({ commentId, userId: user.id });
    } catch (err) {
      console.error(err);
      setError(err);
    }
  };

  return (
    <CommentsContext.Provider
      value={{
        comments,
        error,
        postComment,
        toggleCommentLike,
        fetchComments,
      }}
    >
      {children}
    </CommentsContext.Provider>
  );
};

export const useComments = () => useContext(CommentsContext);
