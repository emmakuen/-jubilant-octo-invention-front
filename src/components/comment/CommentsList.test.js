import { render, screen, waitFor, within } from "@testing-library/react";
import "@testing-library/jest-dom";
import { MemoryRouter } from "react-router-dom";
import CommentsList from "./CommentsList";
import COMMENTS from "./__data__/comments";
import { CommentsContext } from "../../contexts/useComments";
import * as socketConnection from "../../socket/socketConnection";
import userEvent from "@testing-library/user-event";

const USER = {
  id: 1,
  email: "robhope@example.com",
  name: "Rob Hope",
  profilePicUrl: "https://randomuser.me/api/portraits/men/75.jpg",
};

jest.mock("../../socket/socketConnection", () => {
  return {
    listenForToggleUpvote: jest.fn((data) => data),
    removeToggleUpvoteListener: jest.fn(),
  };
});

const toggleCommentLike = jest.fn();
const postComment = jest.fn();

const Wrapper = ({ children }) => {
  return (
    <CommentsContext.Provider value={{ toggleCommentLike, postComment }}>
      <MemoryRouter>{children}</MemoryRouter>
    </CommentsContext.Provider>
  );
};

const renderWithWrapper = (ui, options) => {
  return render(ui, { wrapper: Wrapper, ...options });
};

describe("CommentsList", () => {
  beforeEach(() => {
    jest.restoreAllMocks();
  });

  test("render list of comments correctly", () => {
    renderWithWrapper(<CommentsList comments={COMMENTS} user={USER} />);

    const authorNames = screen
      .getAllByTestId("author-name")
      .map((element) => element.textContent);

    const replyAuthorNames = screen
      .getAllByTestId("reply-author-name")
      .map((element) => element.textContent);
    const commentContainer = screen.getByTestId("comment-list");
    expect(authorNames).toMatchSnapshot();
    expect(replyAuthorNames).toMatchSnapshot();
    expect(commentContainer).toMatchSnapshot();
  });

  test("listen for upvote event and toggle upvote on button click", async () => {
    renderWithWrapper(<CommentsList comments={[COMMENTS[0]]} user={USER} />);

    expect(socketConnection.listenForToggleUpvote).toBeCalled();
    const upvoteButton = screen.getByRole("button", { name: /â–² upvote/i });
    userEvent.click(upvoteButton);

    await waitFor(() => {
      expect(toggleCommentLike).toHaveBeenCalled();
    });
  });

  test("dialog opens when reply button is clicked and closes when cancel is clicked", async () => {
    renderWithWrapper(<CommentsList comments={[COMMENTS[0]]} user={USER} />);

    const replyButton = screen.getByRole("button", { name: /reply/i });
    userEvent.click(replyButton);
    await waitFor(() => {
      expect(screen.getByTestId("dialog")).toBeInTheDocument();
    });

    const cancelButton = screen.getByRole("button", { name: /cancel/i });
    userEvent.click(cancelButton);
    await waitFor(() => {
      expect(screen.queryByTestId("dialog")).not.toBeInTheDocument();
    });
  });

  test("dialog posts reply and closes when post button is clicked", async () => {
    renderWithWrapper(<CommentsList comments={[COMMENTS[0]]} user={USER} />);

    const replyButton = screen.getByRole("button", { name: /reply/i });
    userEvent.click(replyButton);

    // initially, expect post button to be disabled when input is empty
    const postButton = screen.getByRole("button", { name: /post/i });
    expect(postButton).toBeDisabled();

    // after typing input and clicking post button,
    // expect comment to be posted and dialog to be closed
    const dialog = screen.getByTestId("dialog");
    const input = within(dialog).getByRole("textbox");
    const content = "hello";
    userEvent.type(input, content);
    userEvent.click(postButton);
    await waitFor(() => expect(dialog).not.toBeInTheDocument());
    expect(postComment).toHaveBeenCalledWith({
      parentId: COMMENTS[0].id,
      content,
    });
  });
});
