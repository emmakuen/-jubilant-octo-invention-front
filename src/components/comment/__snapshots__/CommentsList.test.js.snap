// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommentsList render list of comments correctly 1`] = `
Array [
  "Cameron Lawrence",
  "Sophie Brecht",
]
`;

exports[`CommentsList render list of comments correctly 2`] = `
Array [
  "James",
]
`;

exports[`CommentsList render list of comments correctly 3`] = `
<div
  data-testid="comment-list"
>
  <div>
    <div
      class="css-zodmwq"
    >
      <div
        class="css-5zthi9"
      >
        <a
          aria-label="See user profiles"
          href="/"
          role="button"
        >
          <img
            alt="Cameron Lawrence"
            class="css-oovt6k"
            src="https://randomuser.me/api/portraits/women/70.jpg"
          />
        </a>
      </div>
      <p
        class="css-1wq22vh"
      >
        <a
          aria-label="See user profiles"
          href="/"
          role="button"
          tabindex="0"
        >
          <span
            class="css-wpbqzu"
            data-testid="author-name"
          >
            Cameron Lawrence
          </span>
        </a>
        <span
          class="css-1vkt2zm"
        >
          4 weeks ago
        </span>
      </p>
      <p
        class="css-v7n70w"
      >
        Love the native memberships and the zipless themes, I was just asked by a friend about options for a new site, and I think I know what I'll be recommending then...
      </p>
      <div
        class="css-1jlz688"
      >
        <button
          class="css-1f6sehp"
        >
          ▲ Upvote
        </button>
        <button
          class="css-1f6sehp"
        >
          Reply
        </button>
      </div>
    </div>
  </div>
  <div>
    <div
      class="css-m8nkj0"
    >
      <div
        class="css-5zthi9"
      >
        <a
          aria-label="See user profiles"
          href="/"
          role="button"
        >
          <img
            alt="Sophie Brecht"
            class="css-oovt6k"
            src="https://randomuser.me/api/portraits/women/75.jpg"
          />
        </a>
      </div>
      <p
        class="css-1wq22vh"
      >
        <a
          aria-label="See user profiles"
          href="/"
          role="button"
          tabindex="0"
        >
          <span
            class="css-wpbqzu"
            data-testid="author-name"
          >
            Sophie Brecht
          </span>
        </a>
        <span
          class="css-1vkt2zm"
        >
          One week ago
        </span>
      </p>
      <p
        class="css-v7n70w"
      >
        Switched our blog from Hubspot to Ghost a year ago -- turned out to be a great decision. Looking forward t othis update... The in-platform analytics look especially delicious. :)
      </p>
      <div
        class="css-1jlz688"
      >
        <button
          class="css-1f6sehp"
        >
          ▲ Upvote
        </button>
        <button
          class="css-1f6sehp"
        >
          Reply
        </button>
      </div>
    </div>
    <div
      class="css-h1ux1i"
    >
      <div
        class="css-5zthi9"
      >
        <a
          aria-label="See user profiles"
          href="/"
          role="button"
        >
          <img
            alt="James"
            class="css-oovt6k"
            src="https://randomuser.me/api/portraits/men/74.jpg"
          />
        </a>
      </div>
      <p
        class="css-1wq22vh"
      >
        <a
          aria-label="See user profiles"
          href="/"
          role="button"
          tabindex="0"
        >
          <span
            class="css-wpbqzu"
            data-testid="reply-author-name"
          >
            James
          </span>
        </a>
        <span
          class="css-1vkt2zm"
        >
          One week ago
        </span>
      </p>
      <p
        class="css-v7n70w"
      >
        Thanks Sophie! Last year has been an absolute goldrush for the creator economy. Slowly at first, then all at once. Will be interesting to see how this ecosystem evolves over the next few years.
      </p>
      <div
        class="css-1jlz688"
      >
        <button
          class="css-1f6sehp"
        >
          ▲ Upvote
        </button>
        <button
          class="css-1f6sehp"
        >
          Reply
        </button>
      </div>
    </div>
  </div>
</div>
`;
